### PR TITLE
Remove reply button on old comments

### LINKED ISSUE #823: Block commenting on old threads
It's pretty ridiculous when people bump up threads from years ago, adding nothing of value to the conversation.

### PR DESCRIPTION
**Pull request by int-y1**
fixes #823

* remove reply button when comment is too old and user doesn't have `judge.change_comment`.
* users can now reply to their own comment (previously, the backend allowed it, but the frontend didn't show the reply button)

Capture

### CODE REVIEW

**Comment by Riolku** [comment]
Do other people have opinions on this being a pure UI change?

**Comment by int-y1** [comment]
> It looks to be a trivial change in the backend, so let's do it instead of letting people deliberately necropost to "show their 1337 h4x0r skills".

done

File: dmoj/settings.py [hunk 1]
```diff
@@ -85,6 +85,7 @@
 DMOJ_USER_DATA_CACHE = ''
 DMOJ_USER_DATA_DOWNLOAD_RATELIMIT = datetime.timedelta(days=1)
 DMOJ_COMMENT_VOTE_HIDE_THRESHOLD = -5
+DMOJ_COMMENT_REPLY_TIMEFRAME = datetime.timedelta(days=1000)
```
Comment (Riolku) [inline dmoj/settings.py:88]: Maybe 365 is better because it's less arbitrary.

File: judge/comments.py [hunk 2]
```diff
@@ -78,10 +79,14 @@ def post(self, request, *args, **kwargs):
...
+                return HttpResponseBadRequest()
+            try:
+                parent_comment = Comment.objects.get(hidden=False, id=parent, page=page)
+            except Comment.DoesNotExist:
                 return HttpResponseNotFound()
-            else:
-                if not Comment.objects.filter(hidden=False, id=parent, page=page).exists():
-                    return HttpResponseNotFound()
+            if not (self.request.user.has_perm('judge.change_comment') or
+                    parent_comment.time > timezone.now() - settings.DMOJ_COMMENT_REPLY_TIMEFRAME):
```
Comment (Riolku) [inline judge/comments.py:88]: I might be the only one, but I feel like

```
if timezone.now() - parent_comment.time > settings.DMOJ_COMMENT_REPLY_TIMEFRAME and not self.request.user.has_perm('judge.change_comment'):
```

Is the more readable way of writing this.

File: templates/comments/list.html [hunk 3]
```diff
@@ -66,39 +66,32 @@ <h2><i style="padding-right: 0.3em" class="fa fa-comments"></i>{{ _('Comments')
                                 {% else %}
                                     <span class="comment-edits"></span>
                                 {% endif %}
-                                    <a href="#comment-{{ node.id }}" title="{{ _('Link') }}" class="comment-link">
-                                    <i class="fa fa-link fa-fw"></i>
-                                </a>
+                                    <a href="#comment-{{ node.id }}" title="{{ _('Link') }}"
+                                       class="comment-link"><i class="fa fa-link fa-fw"></i></a>
```
Comment (Riolku) [inline templates/comments/list.html:70]: I think `</a>` should be on its own line.

File: judge/comments.py [hunk 4]
```diff
@@ -118,5 +123,6 @@ def get_context_data(self, **kwargs):
             context['is_new_user'] = not self.request.user.is_staff and not profile.has_any_solves
         context['comment_list'] = queryset
         context['vote_hide_threshold'] = settings.DMOJ_COMMENT_VOTE_HIDE_THRESHOLD
+        context['reply_timeframe'] = timezone.now() - settings.DMOJ_COMMENT_REPLY_TIMEFRAME
```
Comment (Riolku) [inline judge/comments.py:126]: Let's call this something like `can_reply_before`.

**Review by Riolku — COMMENTED** [review]
Overall looks fine.

File: templates/comments/list.html (hunk 3 above)
Comment (int-y1) [reply templates/comments/list.html:70]: `</a>` appears multiple times, and moving each `</a>` into a new line hurts readability

File: judge/comments.py (hunk 4 above)
Comment (int-y1) [reply judge/comments.py:126]: `{% if node.time > can_reply_before %}` is too confusing for me

File: dmoj/settings.py (hunk 1 above)
Comment (int-y1) [reply dmoj/settings.py:88]: i should collect statistics to see if 365 is a good cutoff

File: judge/comments.py (hunk 4 above)
Comment (Riolku) [reply judge/comments.py:126]: what about `reply_cutoff`

Comment (int-y1) [reply judge/comments.py:126]: sure

**Review by kiritofeng — APPROVED** [review]
LGTM, thanks!

**Review by Xyene — APPROVED** [review]
LGTM, thanks!

**Pull request merged**